pageserver/page_api/src/client.rs

```diff
@@ -0,0 +1,234 @@
+        let shard_ascii : AsciiMetadataValue = self.shard_id.clone().try_into().unwrap();
+
+        metadata.insert("neon-tenant-id", tenant_ascii);
+        metadata.insert("neon-timeline-id", timeline_ascii);
+        metadata.insert("neon-shard-id", shard_ascii);
+
+        return (req, self);
+    }
+
+    pub async fn check_rel_exists(
+        mut self,
+        req: model::CheckRelExistsRequest,
+    ) -> Result<model::CheckRelExistsResponse, tonic::Status> {
+
+        let proto_req = proto::CheckRelExistsRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            rel: Some(req.rel.into()),
+        };
+
+        let mut req_new = Request::new(proto_req);
+        (req_new, self) = self.insert_metadata::<proto::CheckRelExistsRequest>(req_new);
+
+        let response = self
+            .client
+            .check_rel_exists(req_new)
+            .await?;
+
+        let proto_resp = response.into_inner();
+        Ok(proto_resp.into())
+    }
+
+    pub async fn get_base_backup(
+        mut self,
+        req: &model::GetBaseBackupRequest,
+    ) -> Result<impl Stream<Item = Result<Bytes, tonic::Status>>, tonic::Status> {
+
+        let proto_req = proto::GetBaseBackupRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            replica: req.replica,
+        };
+
+        let mut req_new = Request::new(proto_req);
+        (req_new, self) = self.insert_metadata::<proto::GetBaseBackupRequest>(req_new);
+
+        let response_stream: Streaming<proto::GetBaseBackupResponseChunk> = self
+            .client
+            .get_base_backup(req_new)
+            .await?
+            .into_inner();
+
+        let domain_stream = response_stream.map(|chunk_res| {
+            chunk_res.and_then(|proto_chunk| {
+                let b: Bytes = proto_chunk.try_into().unwrap();
+                Ok(b)
+            })
+        });
+
+        Ok(domain_stream)
+    }
+
+    pub async fn get_db_size(
+        mut self,
+        req: model::GetDbSizeRequest,
+    ) -> Result<u64, tonic::Status> {
+
+        let proto_req = proto::GetDbSizeRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            db_oid: req.db_oid,
+        };
+
+        let mut req_new = Request::new(proto_req);
+        (req_new, self) = self.insert_metadata::<proto::GetDbSizeRequest>(req_new);
+
+        let response = self.client.get_db_size(req_new).await?;
+        Ok(response.into_inner().into())
+    }
+
+    pub async fn get_pages<ReqSt>(
+        mut self,
+        inbound: ReqSt,
+    ) -> Result<impl Stream<Item = Result<model::GetPageResponse, tonic::Status>> + Send + 'static, tonic::Status>
+    where
+        ReqSt: Stream<Item = model::GetPageRequest> + Send + 'static,
+    {
+        let outbound_proto: Pin<Box<dyn Stream<Item = proto::GetPageRequest> + Send>> =
+            Box::pin(
+                inbound.map(|domain_req| {
+                    domain_req.try_into().unwrap()
```

Comment:
This error handling will be annoying, and is unnecessary. I'm making the conversion infallible in #12115.